📝 update docs (#3181)
mondoohq/cnquery: docs/development.md

)
```

## Providers development best practices

> The more time we spent building providers, the more learn about how to do that better in the future. Here we described learnings that would help you get started with providers development.

Reviewer: ```suggestion
The more time we spend building providers, the more learn how to do better in the future. Here we describe learnings that will help you get started with providers development.
```

### Referencing MQL resources
Often we have a top-level MQL resource, which we want to reference in another top-level resource. 

For example, GCP networks can be retrieved for a project. That is a top-level resource:
```
// GCP Compute Engine
private gcp.project.computeService {
   // Google Compute Engine VPC network in a project
   networks() []gcp.project.computeService.network
}
```

> However, we have a reference to a GCP network in a GCP Compute Address. This allows us to quickly navigate to the network in which an address is created:

Reviewer: ```suggestion
However, we have a reference to a GCP network in a GCP Compute address. This allows us to quickly navigate to the network in which an address is created:
```

```
private gcp.project.computeService.address {
  // Static IP address
  address string

  // Network in which to reserve the address
  network() gcp.project.computeService.network
}
```

> The simple wait to implement the reference would be to call the GCP API every time `gcp.project.computeService.address.network` is executed. However, this means we are going to be doing an excessive amount of API calls when scanning large GCP projects. If we have 10 addresses, this would mean we will do 10 separate API calls to get the network for each of them.

Reviewer: ```suggestion
The simple way to implement the reference would be to call the GCP API every time `gcp.project.computeService.address.network` is executed. However, this would generate an excessive amount of API calls when scanning large GCP projects. If we have 10 addresses, this would mean 10 separate API calls to get the network, one for each of them.
```

> MQL has powerful caching capabilities that would allow to achieve the same end result with a single (or fewer) API calls. 

Reviewer: ```suggestion
MQL has powerful caching capabilities that let us achieve the same end result with a single (or fewer) API calls. 
```

		"projectId": args["projectId"],
	})
	if err != nil {
		return nil, nil, err
	}

   // Cast the resource to the appropriate type
	computeSvc := obj.(*mqlGcpProjectComputeService)
   // List the networks: equivalent to gcp.project.computeService.networks MQL query. This will retrieve all networks in the project and cache them in the MQL cache. Consecutive calls to this will retrieve the data from cache and will not execute any API calls.
	networks := computeSvc.GetNetworks()
	if networks.Error != nil {
		return nil, nil, networks.Error
	}

   // Filter the networks in memory by comparing them with the input arguments
	for _, n := range networks.Data {
		network := n.(*mqlGcpProjectComputeServiceNetwork)
		name := network.GetName()
		if name.Error != nil {
			return nil, nil, name.Error
		}
		projectId := network.GetProjectId()
		if projectId.Error != nil {
			return nil, nil, projectId.Error
		}

      // return the resource if found
		if name.Data == args["name"].Value && projectId.Data == args["projectId"].Value {
			return args, network, nil
		}
	}
	return nil, nil, fmt.Errorf("not found")
}
```

> Then, we implement the function for retrieving the network for a GCP compute address as follows:

Reviewer: ```suggestion
Then, we implement the function for retrieving the network for a GCP compute address:
```
